### PR TITLE
Fix PathUtil handling of spaces in file names

### PR DESCRIPTION
**Pull request by mdonatas**
## Proposed changes

- Currently PathUtil strips leading spaces in file names resulting in "file not found" messages when viewing files in FormCommit
- Encoding file name before giving it to Uri fixes this

## Screenshots
image

## Test methodology

- Unit tests + manually

## Test environment(s) 

- Git Extensions 33.33.33
- Build 2c32e7e3ce43586ad1fde3b7562b6476463c575d
- Git 2.31.1.windows.1
- Microsoft Windows NT 10.0.19042.0
- .NET 5.0.6
- DPI 192dpi (200% scaling)

----

:black_nib: I contribute this code under [The Developer Certificate of Origin](../blob/master/contributors.txt).

### CODE REVIEW

**Review by RussKie — APPROVED** [review]

File: UnitTests/GitCommands.Tests/Helpers/PathUtilTest.cs [hunk 1]
```diff
@@ -185,6 +185,7 @@ public void Resolve(string path, string expected)
         }
 
         [TestCase(@"C:\WORK\", @"GitExtensions\", @"C:\WORK\GitExtensions\")]
+        [TestCase(@"C:\WORK\", @" file.txt", @"C:\WORK\ file.txt")]
```
Comment (gerhardol) [inline UnitTests/GitCommands.Tests/Helpers/PathUtilTest.cs:188]: Can you please add a testcase for trailing space as well as space before/after /?
(could be in this string)
I
```suggestion
        [TestCase(@"C:\WORK\", @" file.txt ", @"C:\WORK\ file.txt ")]
        [TestCase(@"C:\WORK\", @" \ dir \ file.txt ", @"C:\WORK\ \ dir \ file.txt ")]
```

**Review by gerhardol — APPROVED** [review]

**Comment by gerhardol** [comment]
If this is broken in 3.5 and working in 3.4, a separate PR should be submitted to 3.5

**Review by mstv — COMMENTED** [review]
The `FullPathResolverTests.Resolve_should_return_full_path` fails for the almost-64K path:
```
System.UriFormatException : Invalid URI: The Uri string is too long.
   at System.Uri.CreateThis(String uri, Boolean dontEscape, UriKind uriKind)
   at System.Uri.CreateUri(Uri baseUri, String relativeUri, Boolean dontEscape)
   at System.Uri..ctor(Uri baseUri, String relativeUri)
   at GitCommands.PathUtil.ResolveRelativePath(String path, String relativePath) in C:\projects\gitextensions\GitCommands\PathUtil.cs:line 153
   at GitCommands.PathUtil.Resolve(String path, String relativePath) in C:\projects\gitextensions\GitCommands\PathUtil.cs:line 121
   at GitCommands.FullPathResolver.Resolve(String path) in C:\projects\gitextensions\GitCommands\FullPathResolver.cs:line 65
   at GitCommandsTests.FullPathResolverTests.Resolve_should_return_full_path(String path) in C:\projects\gitextensions\UnitTests\GitCommands.Tests\FullPathResolverTests.cs:line 48
```
What would be an acceptable limit of the maximum path length supported?

**Comment by RussKie** [comment]
260 chars ;)

**Comment by gerhardol** [comment]
> What would be an acceptable limit of the maximum path length supported?

Is this PR changing the behavior for relative paths?
This is something that has been reorted on, both when GE internal handling wxceeds 260 chars and when the actual path is over 260. If this is the first, it should be addressed.

File: UnitTests/GitCommands.Tests/Helpers/PathUtilTest.cs (hunk 1 above)
Comment (mdonatas) [reply UnitTests/GitCommands.Tests/Helpers/PathUtilTest.cs]: Changed it to `@" file .txt "`. From my limited testing it seems that a space at the end of a file name is not possible in Windows but is possible under Linux.

**Comment by mdonatas** [comment]
> 
> 
> > What would be an acceptable limit of the maximum path length supported?
> 
> Is this PR changing the behavior for relative paths?
> This is something that has been reorted on, both when GE internal handling wxceeds 260 chars and when the actual path is over 260. If this is the first, it should be addressed.

I don't think that it does.. the limit for path length (after some googling) seems to be a puny 260 chars for Windows and 4096 for Linux so a test-case with 64KB worth of path is an overkill :)
Apart from using relative paths which after encoding + base path would exceed 64KB external behavior doesn't change.

**Comment by mdonatas** [comment]
> 
> 
> The `FullPathResolverTests.Resolve_should_return_full_path` fails for the almost-64K path:
> 
> ```
> System.UriFormatException : Invalid URI: The Uri string is too long.
>    at System.Uri.CreateThis(String uri, Boolean dontEscape, UriKind uriKind)
>    at System.Uri.CreateUri(Uri baseUri, String relativeUri, Boolean dontEscape)
>    at System.Uri..ctor(Uri baseUri, String relativeUri)
>    at GitCommands.PathUtil.ResolveRelativePath(String path, String relativePath) in C:\projects\gitextensions\GitCommands\PathUtil.cs:line 153
>    at GitCommands.PathUtil.Resolve(String path, String relativePath) in C:\projects\gitextensions\GitCommands\PathUtil.cs:line 121
>    at GitCommands.FullPathResolver.Resolve(String path) in C:\projects\gitextensions\GitCommands\FullPathResolver.cs:line 65
>    at GitCommandsTests.FullPathResolverTests.Resolve_should_return_full_path(String path) in C:\projects\gitextensions\UnitTests\GitCommands.Tests\FullPathResolverTests.cs:line 48
> ```
> 
> What would be an acceptable limit of the maximum path length supported?

It seems that Linux has a limit of 4096 so going much beyond that is not necessary but since I don't know how to choose this arbitrary limit I simply shortened that test-case to still fit the resulting path into 64KB

**Review by mstv — APPROVED** [review]
Did the second proposed test case with spaces in folder names fail?

**Comment by mdonatas** [comment]
> 
> 
> Did the second proposed test case with spaces in folder names fail?

Uff, I've completely misread the proposed diff and took a second line as something which has already existed so didn't even read it :D now that I see it, the test case passes just fine but I am more hesitant to include it as a `requirement` on the `PathUtil`. Just look at what it takes to work with such folders https://superuser.com/questions/915197/how-do-i-create-a-directory-with-a-trailing-space I bet many other things would break left and right before this ever comes to execute. Ideas as to whether this requirement on behavior should really be there?

**Comment by RussKie** [comment]
Please send a matching PR to 3.5 branch

**Pull request merged**